### PR TITLE
Printing static code analysis errors in log

### PR DESCRIPTION
**Pull request by tomasbjerre**
The build log may look like:

```bash
[INFO] 
[INFO] --- violations-maven-plugin:1.24:violations (default) @ generic-webhook-trigger ---
[INFO] No references specified, will not report violations in diff
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  55.612 s
[INFO] Finished at: 2019-10-05T10:38:50+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal se.bjurr.violations:violations-maven-plugin:1.24:violations (default) on project generic-webhook-trigger: Too many violations found, max is 0 but found 2
[ERROR] org/jenkinsci/plugins/gwt/GenericTrigger.java
[ERROR] |          |          |          |      |                                   |
[ERROR] | Reporter | Rule     | Severity | Line | Message                           |
[ERROR] |          |          |          |      |                                   |
[ERROR] +----------+----------+----------+------+-----------------------------------+
[ERROR] |          |          |          |      |                                   |
[ERROR] | Spotbugs | RV_RETUR | INFO     | 144  | Return value  of  method  without |
[ERROR] |          | N_VALUE_ |          |      | side effect  is  ignored  <p>This |
[ERROR] |          | IGNORED_ |          |      | code calls a method  and  ignores |
[ERROR] |          | NO_SIDE_ |          |      | the  return  value.  However  our |
[ERROR] |          | EFFECT   |          |      | analysis shows  that  the  method |
[ERROR] |          |          |          |      | (including its implementations in |
[ERROR] |          |          |          |      | subclasses  if  any)   does   not |
[ERROR] |          |          |          |      | produce  any  effect  other  than |
[ERROR] |          |          |          |      | return value. Thus this call  can |
[ERROR] |          |          |          |      | be removed. </p> <p>We are trying |
[ERROR] |          |          |          |      | to reduce the false positives  as |
[ERROR] |          |          |          |      | much as  possible,  but  in  some |
[ERROR] |          |          |          |      | cases  this  warning   might   be |
[ERROR] |          |          |          |      | wrong.   Common    false-positive |
[ERROR] |          |          |          |      | cases   include:</p>   <p>-   The |
[ERROR] |          |          |          |      | method   is   designed   to    be |
[ERROR] |          |          |          |      | overridden  and  produce  a  side |
[ERROR] |          |          |          |      | effect in  other  projects  which |
[ERROR] |          |          |          |      | are  out  of  the  scope  of  the |
[ERROR] |          |          |          |      | analysis.</p> <p>- The method  is |
[ERROR] |          |          |          |      | called  to  trigger   the   class |
[ERROR] |          |          |          |      | loading which  may  have  a  side |
[ERROR] |          |          |          |      | effect.</p> <p>-  The  method  is |
[ERROR] |          |          |          |      | called   just   to    get    some |
[ERROR] |          |          |          |      | exception.</p>  <p>If  you   feel |
[ERROR] |          |          |          |      | that our assumption is incorrect, |
[ERROR] |          |          |          |      | you can use  a  @CheckReturnValue |
[ERROR] |          |          |          |      | annotation to  instruct  FindBugs |
[ERROR] |          |          |          |      | that ignoring the return value of |
[ERROR] |          |          |          |      | this method is acceptable. </p>   |
[ERROR] |          |          |          |      |                                   |
[ERROR] +----------+----------+----------+------+-----------------------------------+
[ERROR] |          |          |          |      |                                   |
[ERROR] | Spotbugs | NP_NULL_ | ERROR    | 140  | Possible null pointer dereference |
[ERROR] |          | ON_SOME_ |          |      | <p>  There   is   a   branch   of |
[ERROR] |          | PATH     |          |      | statement      that,       <em>if |
[ERROR] |          |          |          |      | executed,</em> guarantees that  a |
[ERROR] |          |          |          |      | null value will be  dereferenced, |
[ERROR] |          |          |          |      | which    would     generate     a |
[ERROR] |          |          |          |      | <code>NullPointerException</code> |
[ERROR] |          |          |          |      | when the  code  is  executed.  Of |
[ERROR] |          |          |          |      | course, the problem might be that |
[ERROR] |          |          |          |      | the  branch   or   statement   is |
[ERROR] |          |          |          |      | infeasible  and  that  the   null |
[ERROR] |          |          |          |      | pointer exception can't  ever  be |
[ERROR] |          |          |          |      | executed; deciding that is beyond |
[ERROR] |          |          |          |      | the ability of FindBugs. </p>     |
[ERROR] |          |          |          |      |                                   |
[ERROR] +----------+----------+----------+------+-----------------------------------+
[ERROR] Summary of org/jenkinsci/plugins/gwt/GenericTrigger.java
[ERROR] |          |      |      |       |       |
[ERROR] | Reporter | INFO | WARN | ERROR | Total |
[ERROR] |          |      |      |       |       |
[ERROR] +----------+------+------+-------+-------+
[ERROR] |          |      |      |       |       |
[ERROR] | Spotbugs | 1    | 0    | 1     | 2     |
[ERROR] |          |      |      |       |       |
[ERROR] +----------+------+------+-------+-------+
[ERROR] |          |      |      |       |       |
[ERROR] |          | 1    | 0    | 1     | 2     |
[ERROR] |          |      |      |       |       |
[ERROR] +----------+------+------+-------+-------+
[ERROR] 
[ERROR] 
[ERROR] Summary
[ERROR] |          |      |      |       |       |
[ERROR] | Reporter | INFO | WARN | ERROR | Total |
[ERROR] |          |      |      |       |       |
[ERROR] +----------+------+------+-------+-------+
[ERROR] |          |      |      |       |       |
[ERROR] | Spotbugs | 1    | 0    | 1     | 2     |
[ERROR] |          |      |      |       |       |
[ERROR] +----------+------+------+-------+-------+
[ERROR] |          |      |      |       |       |
[ERROR] |          | 1    | 0    | 1     | 2     |
[ERROR] |          |      |      |       |       |
[ERROR] +----------+------+------+-------+-------+
```

### CODE REVIEW

**Comment by oleg-nenashev** [comment]
Would be nice to have several downstream pull requests demonstrating the enhancement

File: pom.xml [hunk 1]
```diff
@@ -65,19 +65,24 @@
...
          Hint: SpotBugs considers some real NPE risks in Jenkins as low-priority issues, it is recommended to enable it in plugins.
       -->
     <spotbugs.threshold>${findbugs.threshold}</spotbugs.threshold>
     <!-- Defines a SpotBugs effort. Use "Max" to maximize the scan depth -->
     <spotbugs.effort>${findbugs.effort}</spotbugs.effort>
-    
+    <!--
+      Maximum allowed number of violations found with static code analysis, or else fail the build. 
+      It is strongly encouraged to leave it as 0. Use false with care only in transient situations.
+    -->
```
Comment (oleg-nenashev) [inline pom.xml]: It causes incompatibilities for those who set up the 'failOnError' flag. We need to alter max violations if it is set, e.g. via profiles

Comment (tomasbjerre) [reply pom.xml]: Corrected.
`mvn install -Dspotbugs.failOnError=false` allows 999999 violations.
`mvn install -Dspotbugs.failOnError=true` allows 0 violations.

**Comment by tomasbjerre** [comment]
> Would be nice to have several downstream pull requests demonstrating the enhancement

https://github.com/jenkinsci/generic-webhook-trigger-plugin/pull/137

**Comment by tomasbjerre** [comment]
@oleg-nenashev I added profiles to the parent. But I when I try it out in https://github.com/jenkinsci/generic-webhook-trigger-plugin/pull/137 they are not getting activated. I use `failOnError = false` but still it does fail. `mvn help:active-profiles` does not show my profiles being activated. Not really sure why. Will continue investigate that, any help is welcome =)

**Comment by casz** [comment]
@tomasbjerre force pushing makes it harder to reivew 😓 If you wish for the commit to be squashed it can be done so.

**Comment by daniel-beck** [comment]
@tomasbjerre Thanks for this proposal! Having to search what the stupid error codes mean is annoying, and this looks pretty helpful in that case.

However, beyond concerns like James' around the functionality and usefulness itself, this looks pretty much unproven: This has no known uses on mvnrepository.com, no stars or forks on GitHub, and the only notable uses GitHub identifies are plugins and other components by its own maintainer.

Which means to me that, so far, the plugin basically just had to deal with issues its maintainer threw at it, and may well have problems when exposed to the wider plugin ecosystem. I realize this is kind of a chicken-and-egg problem for Tomas as a maintainer, but I'd really like to see something be more battle-tested before we roll it out _this_ widely.

----

I would propose you open a discussion on the dev list asking for plugin maintainers' support for this change, as ultimately it's plugin maintainers who end up using this (and having to deal with the resulting problems).

And finally, even if there's no strong support, individual plugin maintainer can always choose to use this -- they just wouldn't get it through the parent POM.

**Comment by casz** [comment]
There is only one way to get it battle tested :)
I'd say go for it, people can either disable the feature or we can revert the addition if it becomes a huge problem.

**Comment by jtnord** [comment]
> Jenkins already has a lovely way to report these anyway - all plugins can enable these reports and all users can view them (including trends today) whn the plugin is built on ci.jenkins.io

just seen @uhafner's analysis reports where never installed or have been uninstalled on the server :-o

**Comment by uhafner** [comment]
They are still installed, e.g. see https://ci.jenkins.io/job/Plugins/job/analysis-model/job/master/718/spotbugs/

**Comment by olivergondza** [comment]
I find it safer to get this "battle tested" on handfull of the plugins/components first and only promote it to parent-pom for everyone after positive reception.

**Comment by jtnord** [comment]
pipeline-library#121

**Pull request closed without merging**